### ISSUE TITLE
[Rust track] option to let the tests run longer on some exercises?

### ISSUE THREAD

**Issue by ftt** [exercism/exercism#6238]
Hi,

In the Rust track, there is an exercise called "Alphametics" which does take a long time to run if all the tests are being executed. On my computer it takes ~20 seconds, but it's not even the worst time I've seen when sampling some other (perfectly correct) community solutions. My solution times out on the server, obviously.

I can be very much mistaken, but the nature of the exercise is such that it's not easy to speed up by choosing a different approach. If that is not so, please suggest a hint in the exercise. If that is so, please consider increasing the test execution timeout.

**Comment by SleeplessByte** [comment]
Hi there! Unfortunately we're already at the maximum timeout on that infrastructure.

@exercism/rust you may want to consider marking exercises or specific tests as slow/optional like Python and JavaScript do for these that frequently time-out or add further hints.

**Comment by ftt** [comment]
Out of curiosity, what is the timeout?

**Comment by NobbZ** [comment]
@ftt IIRC it is 10 seconds, which also includes the build, not just the test runs.

**Comment by petertseng** [comment]
Another thing we could do is to have the test runner include the `--release` flag when testing specifically `alphametics` (and potentially `palindrome-products`)

**Comment by SleeplessByte** [comment]
Will that make it significantly shorter to run @petertseng ?

**Comment by coriolinus** [comment]
Yes, certain exercises are substantially faster to test in release mode, depending on the student implementation. We handle CI for those by checking for a flag file, `.meta/test-in-release-mode`. I don't know if the online editor has enough context to check for the presence of that file.

**Comment by SleeplessByte** [comment]
We started using the `custom` key in `config.json` for similar stuff. This you could do too!

e.g
```
{
  ...
  "custom": {
     "test-in-release-mode": true
  }
}
```

You receive this file in test-runner (and analyzer), so you can read this out.

**Comment by ftt** [comment]
So, is this something I can use right now? If so, can you please provide more details on how to do that? I couldn't find anything in the docs for Exercism in general or for the Rust track in particular.

**Comment by ftt** [comment]
OK, I am confused. "Alphametics" _does_ have this marker file: https://github.com/exercism/rust/blob/main/exercises/practice/alphametics/.meta/test-in-release-mode. And at least locally, running tests in the release mode helps tremendously (<2 seconds for building and testing together). I don't have any dependencies to pull in my Cargo.toml, so why would it be so slow?

**Comment by coriolinus** [comment]
1. While track CI tests alphametics in release mode against the example solution, the test runner which tests student code doesn't yet handle that marker file (and in fact can't access it, hence the discussion above of a custom flag in the exercise configuration).
2. Even once we've figured out a scalable solution for the marker, it's unsurprising if the machine resources allocated for a test run are less powerful than a home computer. I'd be unsurprised to discover that student test runs still sometimes time out given the release mode marker.

**Comment by ftt** [comment]
So… I am sorry if I am being annoying, but I don't really understand what you are telling me, most likely because I don't have the context you have. So, the fact that example solution is run in a different manner from the student's solution, is this a problem with the Rust track specifically or with how exercism works in general?

**Comment by coriolinus** [comment]
It's hard to generalize across all of Exercism because most tracks have their own track-specific infrastructure set up. As far as I am aware, no track uses the same test runners that students use for its CI jobs. However, it's possible that I'm just not aware of the exception.

**Comment by ftt** [comment]
This can be solved with providing a Cargo.toml with this section added:
```
[profile.test]
opt-level = 3
debug = false```

Should it be documented somewhere? In the exercise's hints? In https://exercism.org/docs/tracks/rust/tests?

**Comment by iHiD** [comment]
@ftt What do these options do? They feel like quite important things to have for all students if they make things run faster! (cc @ErikSchierboom)

**Comment by ErikSchierboom** [comment]
So it looks like the custom key is now being used in the appropriate exercises (https://github.com/exercism/rust/commit/b67b91ac4a910d08bdc4f3dc029c6de285a150bd), but that the test runner has not yet been updated (https://github.com/exercism/rust-test-runner/search?q=test-in-release-mode). There was a PR to update the test runner, but that has been closed: https://github.com/exercism/rust-test-runner/pull/52 I've added my thoughts there.
